blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md

```diff
@@ -0,0 +1,127 @@
+purposes can lead to a maintenance burden. 
+
+First iteration of `Transaction` class uses `setters` to set the data. 
+Future iteration may switch to using fluent API instead of `setters`.
+
+:include-java: com/example/portfolio/ProfitCalculatorWithoutEncapsulationFluentTest.java {title: "fluent API", entry: "createTransaction", bodyOnly: true, removeReturn: true}
+
+From the `ProfitCalculator`'s perspective the way `Transaction` instances are created is irrelevant as it doesn't affect the calculation logic.
+But refactoring `Transaction` will break `ProfitCalculatorTest`. 
+
+`Transaction` is going to be used in other tests. As the number of tests that use `Transaction` instances increases you will be more and more reluctant
+to do refactoring. All your tests that create `Transaction` instances are **brittle** tests now.
+
+If you ever want to refactor `Transaction` class you will have two choices:
+* Refactor and waste time fixing tests.
+* Convince yourself that `Transaction` is good as it is. 
+
+Neither is a good choice. 
+
+> I want you to have a way to protect your future self by encapsulating test input preparation
+
+# Test Input Initial Encapsulation
+
+Let's extract `Transaction` creation code into `TestTransactions`
+
+:include-file: com/example/portfolio/ProfitCalculatorWithBasicEncapsulationTest.java {title: "ProfitCalculatorTest.java with initial encapsulation"}
+
+:include-java: com/example/portfolio/TestTransactions.java {title: "TestTransactions.createTransaction", entry: "createTransaction"}
+
+Less verbosity. A good start.
+
+Question: Can you still spot the problem?
+
+```spoiler {title: "tap me to see remaining problems"}
+* As we add more properties to `Transaction` class, some of the tests may need to set *extra transaction properties*. 
+* Additionally some of the properties will be completely *irrelevant to the business logic* under a test, yet you will be forced to set them anyway.
+```
+
+# Test Input Encapsulation With Webtau Table Data   
+
+I was dealing with test problems like above for years and eventually came up with a solution that fits my needs.
+I hope it will fit your needs as well. The solution is to use flexible data structure like [Webtau TableData](https://twosigma.github.io/webtau/guide/reference/table-data)
+
+:include-file: com/example/portfolio/ProfitCalculatorWithTableDataTest.java {title: "ProfitCalculatorTest.java with TableData encapsulation", commentsType: "inline"}
+
+Note: [Webtau](https://github.com/twosigma/webtau) open source project started as my answer to common testing problems I encountered. 
+While I am not working at [Two Sigma](https://www.twosigma.com/) anymore I am still contributing and using it on a regular basis.     
+
+:include-java: com/example/portfolio/TestTransactions.java {title: "TestTransactions.createTransactions", commentsType: "inline", entries: ["createTransactions", "genTransactionId"]}
+
+Notice how `createTransactions` defaults values when they are not present?
```

Comment:
I would change the sentence like this for clarity - 
...tests that don't need, say, `id` or `lot` are free to ignore them.